### DISCUSSION TITLE
I need help with a vercel deployment.

### DISCUSSION THREAD

**Discussion by snake-py** [vercel/vercel#8123] [Help] [answered]
I have created an API route that runs with MathJax-node but it seems that vercel does not find the files. When I am building locally everything works fine. Anyone an idea why this could be?

```ts
import type { NextApiRequest, NextApiResponse } from 'next';
const mjAPI = require('mathjax-node');

type resData = {
  svg: {
    input: string;
    svg: string;
  };
};

type reqData = {
  input: string;
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<resData>
) {
  const { input } = req.query as reqData;
  const mathString = input as string;
  const svg = await mjAPI.typeset({
    math: mathString,
    format: 'TeX',
    svg: true,
    color: '#000',
  });
  res.status(200).json({ svg });
}
```

the Error Log

```
e_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
    at HTMLScriptElementImpl.dispatchEvent (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
    at /var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:42:17
    at Object.check (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:76:11)
    at ReadStream.<anonymous> (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:95:12)
    at ReadStream.wrappedEnqueued (/var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:253:16)
    at ReadStream.emit (node:events:527:28) TypeError: Cannot read properties of undefined (reading 'resetGlyphs')
    at /var/task/node_modules/mathjax-node/lib/main.js:521:35
    at Function.execute (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:243:26)
    at cb (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:225:59)
    at Object.Execute (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:426:36)
    at Object.ExecuteHooks (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:604:31)
    at Function.execute (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:243:26)
    at cb (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:225:59)
    at Object.Execute (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:426:36)
    at Object.Post (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:537:37)
    at Function.execute (file:///var/task/node_modules/mathjax/unpacked/MathJax.js:243:26)
2022-07-09T20:00:40.850Z	abf3e37a-e3a7-42b7-a130-2b62375b0b9e	ERROR	Error: Could not load script: "file:///var/task/node_modules/mathjax/unpacked/extensions/toMathML.js?V=2.7.9"
    at /var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:44:23
    at Object.check (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:76:11)
    at ReadStream.<anonymous> (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:95:12)
    at ReadStream.wrappedEnqueued (/var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:253:16)
    at ReadStream.emit (node:events:527:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) [Error: ENOENT: no such file or directory, open '/var/task/node_modules/mathjax/unpacked/extensions/toMathML.js'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/var/task/node_modules/mathjax/unpacked/extensions/toMathML.js'
}
2022-07-09T20:00:40.853Z	abf3e37a-e3a7-42b7-a130-2b62375b0b9e	ERROR	File load error: file:///var/task/node_modules/mathjax/unpacked/jax/element/mml/jax.js
2022-07-09T20:00:40.854Z	abf3e37a-e3a7-42b7-a130-2b62375b0b9e	ERROR	Error: Could not load script: "file:///var/task/node_modules/mathjax/unpacked/jax/element/mml/jax.js?V=2.7.9"
    at /var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:44:23
    at Object.check (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:76:11)
    at ReadStream.<anonymous> (/var/task/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:95:12)
    at ReadStream.wrappedEnqueued (/var/task/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:253:16)
    at ReadStream.emit (node:events:527:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) [Error: ENOENT: no such file or directory, open '/var/task/node_modules/mathjax/unpacked/jax/element/mml/jax.js'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/var/task/node_modules/mathjax/unpacked/jax/element/mml/jax.js'
}
2022-07-09T20:00:48.554Z abf3e37a-e3a7-42b7-a130-2b62375b0b9e Task timed out after 10.02 seconds
```

**Comment by amyegan** [comment]
I'm sorry to hear you're having trouble with this deployment. Thanks for sharing the error log and code! If you can also answer the following questions, that will help me get a better understanding of what's happening.
- Have you ever had a successful deployment of this feature in the past, or have you been seeing this error since the beginning?
- Do you have a specific version of `mathjax` pinned in your package.json file, or is it flexible? 
- Does it still work locally if you remove `node_modules` and do a fresh install of all packages?

**Comment by snake-py** [comment]
@amyegan Thank you for you rreply. I honestly ditched the backend l.ib and now I am using the frontend version of it. TO answer your questions anyway

- The deployments were all successful, the error log came when I hit the API route
- I had version 2.1 fixed 
- It worked locally after reinstalling everything. even my local next build worked.

**Comment by AdamowiczPawel** [comment]
Hello,

I have a similar issue; however, I ultimately need to use MathJax with Node.js (using mathjax-node), not in the browser .

I am uploading a React/Node.js app to Vercel. 

On my local machine, everything works fine (also after clear reinstallations). Mathjax-node consistently works.
However, on Vercel, I have not managed to make it work. I am using mathjax-node version 2.1.1. I am receiving numerous errors in the Vercel logs after the fetch, which is using a file that imports mathjax-node. The logs are as follows:

> de/node_modules/mathjax/unpacked/MathJax.js:243:26)
>     at cb (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:225:59)
>     at Object.Execute (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:426:36)
>     at Object.ExecuteHooks (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:604:31)
>     at Function.execute (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:243:26)
>     at cb (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:225:59)
>     at Object.Execute (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:426:36)
>     at Object.Post (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:537:37)
>     at Function.execute (file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/MathJax.js:243:26)
> Error: Could not load script: "file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/extensions/toMathML.js?V=2.7.9"
>     at /var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:44:23
>     at Object.check (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:76:11)
>     at ReadStream.<anonymous> (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:95:12)
>     at ReadStream.wrappedEnqueued (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:253:16)
>     at ReadStream.emit (node:events:517:28)
>     at emitErrorNT (node:internal/streams/destroy:151:8)
>     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
>     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) [Error: ENOENT: no such file or directory, open '/var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/extensions/toMathML.js'] {
>   errno: -2,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: '/var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/extensions/toMathML.js'
> }
> File load error: file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/jax/element/mml/jax.js
> Error: Could not load script: "file:///var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/jax/element/mml/jax.js?V=2.7.9"
>     at /var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:44:23
>     at Object.check (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:76:11)
>     at ReadStream.<anonymous> (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:95:12)
>     at ReadStream.wrappedEnqueued (/var/task/backend/node_modules/mathjax-node/node_modules/jsdom/lib/jsdom/browser/resource-loader.js:253:16)
>     at ReadStream.emit (node:events:517:28)
>     at emitErrorNT (node:internal/streams/destroy:151:8)
>     at emitErrorCloseNT (node:internal/streams/destroy:116:3)
>     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) [Error: ENOENT: no such file or directory, open '/var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/jax/element/mml/jax.js'] {
>   errno: -2,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: '/var/task/backend/node_modules/mathjax-node/node_modules/mathjax/unpacked/jax/element/mml/jax.js'
> }

Part of my code:
```
import sharp from "sharp";
import mj from "mathjax-node";

import { equations } from "./equations.js";


mj.config({
  MathJax: {
    SVG: {
      font: "TeX",
    },
    loader: { load: ["[Contrib]/a11y/sre-full.js"] },
  },
});


export const equationConverter = async (equationKey, variables) => {
  const mathMLExpression = equations[equationKey](...Object.values(variables));

  const result = await new Promise((resolve, reject) => {
    mj.typeset(
      {
        math: mathMLExpression,
        format: "MathML",
        svg: true,
      },
      (typesetResult) => {
        if (typesetResult.errors) {
          reject(typesetResult.errors);
        } else {
          resolve(typesetResult);
        }
      }
    );
  });
```
My only one tip is that VSCode shows when i hover over _import mj from "mathjax-node";_ following information: 

> Could not find a declaration file for module 'mathjax-node'. '/home/pawel/VSCode/Projects/mechCalcApp/00_mechCalcApp/backend/node_modules/mathjax-node/lib/main.js' implicitly has an 'any' type.
>   Try `npm i --save-dev @types/mathjax-node` if it exists or add a new declaration (.d.ts) file containing `declare module 'mathjax-node';`ts(7016)

However, for me it is not clear as in local machine everything works regardless of this information. Also:
_`npm i --save-dev @types/mathjax-node`_ - command does not work in my case
_-new declaration file containing `declare module 'mathjax-node_ - also does not work (or maybe i am doing it wrong)

I also would like to add that i am using JS not TS in my app.

If you have any insights or solutions, I would greatly appreciate your assistance.
Best regards,
Pawel

**Reply by snake-py** [reply]
Is your local node version matching with the one from vercel?

**Reply by AdamowiczPawel** [reply]
Yes i have the same version of mathjax-node on local and on vercel.
Error log from my previous post is from version: 2.1.1
Now i downgrade to version 1.3.0 and still the same result, however error log is shorter and more clear to me:
- local: everything works, 
- vercel: errors (new log below)

> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/jax/input/TeX/config.js
> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/jax/input/MathML/config.js
> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/jax/input/AsciiMath/config.js
> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/jax/output/SVG/config.js
> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/jax/output/CommonHTML/config.js
> File load error: file:///var/task/backend/node_modules/mathjax/unpacked/extensions/toMathML.js
> Error: TypeError: Cannot read properties of undefined (reading 'resetGlyphs')
>     at /var/task/backend/node_modules/mathjax-node/lib/main.js:451:31
>     at Function.execute (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:243:26)
>     at cb (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:225:59)
>     at Object.Execute (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:426:36)
>     at Object.ExecuteHooks (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:604:31)
>     at Function.execute (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:243:26)
>     at cb (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:225:59)
>     at Object.Execute (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:426:36)
>     at Object.Post (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:537:37)
>     at Function.execute (file:///var/task/backend/node_modules/mathjax/unpacked/MathJax.js:243:26)

I see that:
-local 
I have in package.json under dependencies only  "mathjax-node": "^2.1.1". After rm-rf node_modules and package-lock.json and clear reinstall of modules, in node_modules folders i can find: mathjax-node and... mathjax

-vercel
I can not find any folder related to mathjax-node or mathjax. I have double checked package.json and update it with different dependencies (all the time no folders of mathjax appears in vercel)  
- "mathjax-node": "^2.1.1" 
- "mathjax-node": "^1.3.0"
- "mathjax-node": "^1.3.0" and  "mathjax": "^3.2.2"

Log from vercel build:

> Restored build cache
> Running "vercel build"
> Vercel CLI 33.5.3
> WARN! Due to `builds` existing in your configuration file, the Build and Development Settings defined in your Project Settings will not apply. Learn More: https://vercel.link/unused-build-settings
> Installing dependencies...
> npm WARN deprecated domexception@1.0.1: Use your platform's native DOMException instead
> npm WARN deprecated abab@2.0.6: Use your platform's native atob() and btoa() methods instead
> npm WARN deprecated left-pad@1.3.0: use String.prototype.padStart()
> npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
> npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
> 
> added 18 packages, removed 3 packages, and changed 10 packages in 2s
> 
> 56 packages are looking for funding
>    run `npm fund` for details
>  Build Completed in /vercel/output [8s]
>  Deploying outputs...
>  
>  Deployment completed
>  Uploading build cache [34.11 MB]...
>  Build cache uploaded: 1.525s

As i can see that mathjax-node is very old I am also trying  to work with mathjax-full, however without success in node.js environment.

**Reply by snake-py** [reply]
I guess one solution might even be to copy the whole package from the node_folder and ship it with your application.

**Reply by AdamowiczPawel** [reply]
Unfortunatelly, this is not the case.
Folders was my mistake, because at initial commit I by mistake added node_modules. After verification i can see that vercel does not include node_modules in "source tab" if you do not ship any with your app. I made a test and ship "mathjax" and "mathjax-node" folders with app and still without succes (I obtain the same results).

I have created a small react/node app which is fetching SVG from server. 
On server mathjax-node package is transforming mathML equation to SVG and sending to frontend when i can display it. 
On local machine evertyhing works.
On Vercel i receive a lot of errors connected to mathjax-node package

Link to github repo: https://github.com/AdamowiczPawel/mathjax_test

Maybe someone will be able to help based on this repo.

**Reply by AdamowiczPawel** [reply, marked as answer]
I managed to solve the problem with mathjax version 3 package instead of mathjax-node. I paste the code below, maybe it will be useful to someone. Mathjax in version 3 works without errors with vercel.

```
import { mathjax } from "mathjax-full/js/mathjax.js";
import { MathML } from "mathjax-full/js/input/mathml.js";
import { SVG } from "mathjax-full/js/output/svg.js";
import { liteAdaptor } from "mathjax-full/js/adaptors/liteAdaptor.js";
import { RegisterHTMLHandler } from "mathjax-full/js/handlers/html.js";

const adaptor = liteAdaptor();
RegisterHTMLHandler(adaptor);

const mathjax_document = mathjax.document("", {
  InputJax: new MathML(),
  OutputJax: new SVG({ fontCache: "local" }),
});

const mathjax_options = {};

function get_mathjax_svg(math) {
  return new Promise(async (resolve, reject) => {
    try {
      const node = await mathjax_document.convert(math, mathjax_options);
      const svg = adaptor.innerHTML(node);
      resolve(svg);
    } catch (error) {
      reject(error);
    }
  });
}

export const handleGenerateSVGmathjax = async (req, res) => {
  let svgResult;

  try {
    const mathML =
      '<math xmlns="http://www.w3.org/1998/Math/MathML"><mi>x</mi></math>';
    svgResult = await get_mathjax_svg(mathML);
  } catch (error) {
    console.error("Error:", error);
  }
  res.send({ svg: svgResult });
};
```


Best reagards
